[Frontend] Ambry blob metadata cache - 2

snalli opened the pull request:
This patch integrates adds metadata cache to frontend non-blocking
router. We conditionally lookup cache on a RANGE request and use
cached metadata. We delete cached metadata during DELETE request.

snalli commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1375,6 +1471,16 @@ protected void maybeProcessCallbacks() {
+        handleMetadataBlob(null, blobInfo.getUserMetadata(), null);
+        // Do not call updateTtlIfRequired(). If the blob expires, the frontend will find out from backend during the next GET request
+        // TODO: Deserialize payload and do basic validations to check if cached blob properties are correct.
+        // TODO: Should we decrypt ? Probably no.
```
One thing I am not very sure of is what to do if the blobInfo we fetch to validate the cache entry is corrupted or not found. Do we invalidate the cache entry and retry to fetch the entire metadata from backend ? @justinlin-linkedin

justinlin-linkedin commented on ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:
```diff
@@ -591,6 +591,11 @@ public class RouterConfig {
+  public static final String ROUTER_SMALLEST_BLOB_FOR_METADATA_CACHE = "router.smallest.blob.for.metadata.cache";
+  @Config(ROUTER_SMALLEST_BLOB_FOR_METADATA_CACHE)
+  public final long routerSmallestBlobForMetadataCache;
+  long numBytesInOneTb = (long) Math.pow(1024, 4);
```
we can use a static value for this and change the variable name to all upper case. 

And 1 TB, really?

justinlin-linkedin replied:
The question is more about why do we differentiate metadata chunk at different size? I don't see why we would not cache some metadata chunk but cache the other.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
+     */
+    boolean shouldLookupMetadataCache() {
+      boolean result = (blobMetadataCache != null &&
+          options.getBlobOptions.getRange() != null &&
```
I think we can remove the `options.getBlobOptions.getRange() != null`. Even if we are trying to get the whole blob, we can still leverage cache.

justinlin-linkedin replied:
That seems to be the case. But we might have multiple clients issuing get requests.

snalli replied:
Sorry, my reply above was for shouldSaveMetadata. But for shouldLookupMetadataCache, in reply to your comment, we should not look up cache on each GET request. We discussed this in the design doc and Ashish suggested to lookup the cache conditionally for eg. only on range requests. This makes sense since vast majority of blobs are small, most requests are GET for small blobs and will not see much benefits from caching. Looking up the cache on each GET just adds more latency. I have some perf numbers to back it up.

snalli replied:
> That seems to be the case. But we might have multiple clients issuing get requests.

Just saw this. I think we both commented at the same time. Yes, you are right, we can get multiple GET requests for the same blob from multiple clients. I will change the code to leverage that but we need to set some threshold for blob size below which we won't cache.

snalli replied:
I thought more about this. If we save metadata even for every GET request on a blob (subject to some min size), then we may want to look it up for subsequent GETs on the same blob. The problem is that we don't know what the size of a blob is just by looking at the blobID. So then we need to lookup the cache for each GET which just adds extra latency for those blobs whose metadata wasn't cached. I discussed this issue in the design doc and adding latency to GETs that will mostly encounter a cache miss wasn't acceptable.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
+     *  False, otherwise.
+     */
+    boolean shouldLookupMetadataCache() {
+      boolean result = (blobMetadataCache != null &&
```
we don't need a boolean variable, we can jus return this statement.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
+          serverBlobProperties = blobInfo.getBlobProperties();
+          chunkIndexToBuf = new ConcurrentHashMap<>();
+          chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+          successfullyDeserialized = true;
```
hmm, this is tricky, we need to set FirstGetChunk to ChunkState.Complete so we don't have to send requests to server nodes. But that being said, we still have to get chunkMetadataList and set chunkIdIterator.  Check out code from line 1616 to line 1636 and method initializeDataChunks.

Arun-LinkedIn replied:
Hmm, are you going with approach of checking cache in GetBlobOperation (and modifying first request to fetch blobInfo instead of metadata chunk) instead of checking cache in NonBlockingRouter to issue separate GetBlobInfo requests? Still thinking that if we go with latter, we wouldn't need with modifications in other methods such as `getOperationFlag`, `handleBody`, etc :)

snalli replied:
To summarise for others:
1/ Check cache in GetBlobOperation and modify code related to FirstGetChunk
2/ Check cache at the Router level. If cache hit, issue GetBlobInfo followed by GetBlob. If cache miss, just issue GetBlob.
Yes I am going with the first approach. All core changes are within FirstGetChunk class. Modifications to methods getOperationFlag, handleBody aren't too difficult. Second approach may avoid this but may require changes elsewhere. We would just be shifting the changes to another place and also scatter it across classes probably. Also, second approach requires callbacks within callbacks which is confusing.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
+     */
+    boolean shouldSaveMetadata() {
+      boolean result = (blobMetadataCache != null &&
+          options.getBlobOptions.getRange() != null &&
```
same above.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1278,13 +1282,100 @@ private class FirstGetChunk extends GetChunk {
+      if (shouldSaveMetadata()) {
+        BlobMetadata blobMetadata = new BlobMetadata(blobId.toString(), blobInfo, compositeBlobInfo);
+        result = blobMetadataCache.putObject(blobMetadata.getBlobId(), blobMetadata);
+        if (result == true) {
```
if (result) {
}

snalli replied:
actually, no. The result var is the result of inserting an object in the cache and not the result of shouldSaveMetadata. So this code is correct.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1308,8 +1399,13 @@ GetOption getGetOption() {
     MessageFormatFlags getOperationFlag() {
-      return options.getBlobOptions.getOperationType() == GetBlobOptions.OperationType.Data ? MessageFormatFlags.Blob
-          : MessageFormatFlags.All;
+      if (isMetadataCacheHit()) {
+        return MessageFormatFlags.BlobInfo;
+      } else if (options.getBlobOptions.getOperationType() == GetBlobOptions.OperationType.Data) {
+        return MessageFormatFlags.Blob;
+      } else {
+        return MessageFormatFlags.All;
+      }
```
this operationFlag is for requests to server node, it shouldn't have anything to do with cache.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1375,6 +1471,16 @@ protected void maybeProcessCallbacks() {
     @Override
     void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInfo messageInfo)
         throws IOException, MessageFormatException {
+
+      if (isMetadataCacheHit()) {
+        getOptions().ageAtAccessTracker.trackAgeAtAccess(serverBlobProperties.getCreationTimeInMs());
+        handleMetadataBlob(null, blobInfo.getUserMetadata(), null);
+        // Do not call updateTtlIfRequired(). If the blob expires, the frontend will find out from backend during the next GET request
+        // TODO: Deserialize payload and do basic validations to check if cached blob properties are correct.
+        // TODO: Should we decrypt ? Probably no.
+        return;
+      }
+
```
handleBody is the method to call when handling response from server node, it shouldn't have anything to do with cache.

snalli replied:
this is the handleBody for the first chunk - metadata chunk. So if we have a cacheHit, we don't need to do anything below this clause. Datachunks have a different handleBody

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1498,10 +1604,13 @@ public List<CompositeBlobInfo.ChunkMetadata> getChunkMetadataList() {
-      ByteBuf serializedMetadataContent = blobData.content();
+      ByteBuf serializedMetadataContent = null;
       try {
-        compositeBlobInfo =
-            MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent.nioBuffer(), blobIdFactory);
+        if (!isMetadataCacheHit()) {
+          serializedMetadataContent = blobData.content();
+          compositeBlobInfo =
+              MetadataContentSerDe.deserializeMetadataContentRecord(serializedMetadataContent.nioBuffer(), blobIdFactory);
+        }
```
same

snalli replied:
explained above

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1529,7 +1638,7 @@ private void handleMetadataBlob(BlobData blobData, byte[] userMetadata, ByteBuff
         }
         if (!rangeResolutionFailure) {
-          if (options.getChunkIdsOnly || getOperationFlag() == MessageFormatFlags.Blob || encryptionKey == null) {
+          if (options.getChunkIdsOnly || getOperationFlag() == MessageFormatFlags.Blob || isMetadataCacheHit() || encryptionKey == null) {
```
same

snalli replied:
explained above

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java:
```diff
@@ -502,6 +510,7 @@ void initiateChunkDeletesIfAny(final String blobIdStr, final String serviceId,
           deleteRequests.add(new BackgroundDeleteRequest(storeKey, serviceId, quotaChargeCallback));
         }
         initiateBackgroundDeletes(deleteRequests);
+        blobMetadataCache.deleteObject(blobIdStr);
```
it's good to remove it from cache in delete, but I think we should remove it from cache in ttl update as well, because ttl update would change the ttl for metadata. So the blob is outdate after ttl update.

snalli replied:
Yeah we discussed this and Tugrul suggested letting the cache entry remain and let the frontend talk to backend to validate the cache entry during the next RANGE request.

snalli replied:
Actually, this reminds me that if the blob doesn't exist in the backend, we should invalidate its cache entry in the frontend. I need to add that code.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -325,6 +328,7 @@ void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback)
       }
       if (firstChunk.isComplete()) {
         onChunkOperationComplete(firstChunk);
+        firstChunk.saveMetadata();
```
this is not the best place to call saveMetadata. poll method will be invoked by network client every 50 ms, so we would call saveMetadata over and over again. I recommend calling saveMetadata in onChunkOperationComplete method after we check operationCallbackInvoked.

justinlin-linkedin replied:
we have a atomicboolean to safe guard the code to be only executed once.

justinlin-linkedin commented:
The PR looks good to me, but I want to propose a new way here, can we create a new class called CachedFirstGetChunk. This CachedFirsrGetChunk would extend from FirstGetChunk. So in the GetBlobOperation's constructor, we would check if this MetadataChunk is in the cached and create a CachedFirstGetChunk. And of course we would fall back to FirstGetChunk when the MetadataChunk is not cached. What do you think? Can we try it?

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1375,6 +1500,36 @@ protected void maybeProcessCallbacks() {
+        BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+        updateTtlIfRequired(serverBlobProperties, messageInfo);
+        if (serverBlobProperties.toString() != receivedBlobProperties.toString()) {
+          /**
+           * If we find blob properties is in our frontend cache, we request just blobInfo from backend to validate
+           * the cache entry. If cached blob properties does not match received blob properties,
+           * we must delete the cache entry and throw a retriable exception to the user.
+           * The next request will re-fill the cache with a valid entry if the blob stills exists.
+           * We do not need to decrypt anything at this point.
+           */
```
I don't think so.

The blob property, except for ttl, is immutable. So if the blob property from server is different from cache, them it's because the ttl is different, which means client updated ttl before get. This just means that we have to update the ttl in the cache, we don't have to through an error back to client.

snalli replied:
I agree. I update the TTL and then compare the cached and received blob properties. This if-clause is for the rare case when something goes wrong.

snalli commented:
> The PR looks good to me, but I want to propose a new way here, can we create a new class called CachedFirstGetChunk. This CachedFirsrGetChunk would extend from FirstGetChunk. So in the GetBlobOperation's constructor, we would check if this MetadataChunk is in the cached and create a CachedFirstGetChunk. And of course we would fall back to FirstGetChunk when the MetadataChunk is not cached. What do you think? Can we try it?

This is a great idea ! I have updated my PR. The code however isn't entirely encapsulated in CachedFirstChunk.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
+      chunkIndexToBuf = new ConcurrentHashMap<>();
+      chunkIndexToBufWaitingForRelease = new ConcurrentHashMap<>();
+      successfullyDeserialized = true;
+      logger.info(String.format("Metadata cache-hit for blobId = %s", blobId));
```
let's do log trace for this. And logger methods are by default formatting your string. we can do it in this way:
```
logger.trace("Metadata cache-hit for blobId = {}", blobId);
```

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
+    }
+
+    /**
+     * Return the {@link MessageFormatFlags} to associate with the first getBlob chunk operation.
+     * @return {@link MessageFormatFlags#Blob} for {@link GetBlobOptions.OperationType#Data}, or
+     *         {@link MessageFormatFlags#All} by default.
+     */
```
we should update the comments here. Let's point out that for cached metadata chunk, we should request blob info from server to verify the existence of this chunk and update TTL.

snalli replied:
good catch!

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
+       */
+      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      if (!serverBlobProperties.toString().equals(receivedBlobProperties.toString())) {
```
verify the equality of blob properties should be done in a different way. we should implement an `equal` method for BlobProperties.

snalli replied:
got it. I thought of the same first !

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
+         */
+        String reason = "Cached blob property does not match received blob property";
+        deleteMetadata(reason);
+        logger.error(
+            String.format("%s for blobId = %s, cached blob property = %s, received blob property = %s", reason, blobId,
+                serverBlobProperties.toString(), receivedBlobProperties));
```
same, logger.error has builtin formatting.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
+        logger.error(
+            String.format("%s for blobId = %s, cached blob property = %s, received blob property = %s", reason, blobId,
+                serverBlobProperties.toString(), receivedBlobProperties));
+        throw new MessageFormatException(reason, MessageFormatErrorCodes.IO_Error);
```
we shouldn't throw MessageFormatException here.

Let's figure out why there is a difference between cached blob properties and requested blob properties. The only case i can think of is ttl.

Rather than that, I can't think so any other case. 

If we updated ttl  and they are still different, then something is probably very wrong.

We should terminate this operation by setting chunkException. And MessageFormatException is not a good one, we should use RouterException.

```
      BlobProperties receivedBlobProperties = MessageFormatRecord.deserializeBlobProperties(payload);
      updateTtlIfRequired(serverBlobProperties, messageInfo);
      lifeVersion = messageInfo.getLifeVersion;
      if (!serverBlobProperties.equals(receivedBlobProperties)) {
            deleteMetadataInCache(reason);
            setChunkException(new RouterException(reason, RouterErrorCode.UnexpectedInternalError));
            return;
      }
```

justinlin-linkedin replied:
And set `chunkCompleted` to true as well.

snalli replied:
should we update lifeVersion here using `blobInfo.setLifeVersion(lifeVersion);`

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1752,107 @@ private void onInvalidRange(Exception exception) {
+      }
+      getOptions().ageAtAccessTracker.trackAgeAtAccess(serverBlobProperties.getCreationTimeInMs());
+
+      /**
+       * If we reach here, then the cached metadata is valid and already decrypted from a previous GET/RANGE.
+       * We can just use it directly and initiate fetching data chunks.
+       */
+      try {
+        if (options.getBlobOptions.hasBlobSegmentIdx()) {
+          int requestedSegment = options.getBlobOptions.getBlobSegmentIdx();
+          if (requestedSegment < 0 || requestedSegment >= chunkMetadataList.size()) {
+            throw new IllegalArgumentException(
+                "Bad segment number: " + requestedSegment + ", num of keys: " + chunkMetadataList.size());
+          }
+          chunkMetadataList = chunkMetadataList.subList(requestedSegment, requestedSegment + 1);
+          if (options.getBlobOptions.getRange() != null) {
+            resolvedByteRange = options.getBlobOptions.getRange().toResolvedByteRange(chunkMetadataList.get(0).getSize());
+          }
+        } else {
+          if (options.getBlobOptions.getRange() != null) {
+            resolvedByteRange = options.getBlobOptions.getRange().toResolvedByteRange(totalSize);
+            chunkMetadataList = compositeBlobInfo.getStoreKeysInByteRange(resolvedByteRange.getStartOffset(),
+                resolvedByteRange.getEndOffset());
+          }
+        }
+        initializeDataChunks();
+      } catch (IllegalArgumentException e) {
+        onInvalidRange(e);
+      }
+    }
+  }
```
can we make a method for this code so we can reuse them?

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -140,18 +144,93 @@ class GetBlobOperation extends GetOperation {
+    if (shouldSaveMetadata()) {
+      BlobMetadata blobMetadata = new BlobMetadata(blobId.toString(), blobInfo, compositeBlobInfo);
+      put_result = blobMetadataCache.putObject(blobMetadata.getBlobId(), blobMetadata);
+      logger.info(String.format("Saving metadata for blobId = %s, result = %s", blobId, put_result));
```
do logger.trace please. And you don't have to call String.format, logger's methods by default do string formatting for you.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -140,18 +144,93 @@ class GetBlobOperation extends GetOperation {
+   * @return True if metadata was saved successfully, else False.
+   */
+  boolean saveMetadata() {
+    boolean put_result = false;
```
let's rename this variable to `putResult`.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -140,18 +144,93 @@ class GetBlobOperation extends GetOperation {
+    if (blobMetadataCache == null) {
+      return false;
+    }
+    boolean delete_result = blobMetadataCache.deleteObject(blobId.toString());
+    logger.info(String.format("Deleting metadata for blobId = %s, reason = %s, result = %s", blobId, reason, delete_result));
```
same

Arun-LinkedIn commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -201,6 +280,15 @@ void abort(Exception abortCause) {
     }
+    switch(((RouterException) abortCause).getErrorCode()) {
+      /**
+       * If the blob is not found, then delete its metadata from frontend cache.
```
nit: probably use end of line comments ( // If the blob is not found ... ) or block comments (/* * */) for multiple lines since we seem to do that in rest of code?

snalli replied:
agreed, will fix

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -219,6 +316,12 @@ private void onChunkOperationComplete(GetChunk chunk) {
     if (chunk == firstChunk) {
       if (operationCallbackInvoked.compareAndSet(false, true)) {
         if (chunk.chunkBlobId.getBlobDataType() == BlobId.BlobDataType.METADATA) {
+          /**
+           * Save metadata on successful completion. Everything is decrypted at this point.
+           */
```
I realize that you have a lot of comments that are using /** */. we should just use // for multiple lines.

snalli replied:
done

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1643,12 +1746,25 @@ private boolean resolveRange(long totalSize) {
+     * @param dataRangeSize byte range to be used for range resolution
+     * @return {@code true} if range resolution succeeded. {@code false} otherwise
+     */
+    protected boolean resolveRangeProper(long dataRangeSize) {
+      /**
+       * resolveRange is a weird function that returns false on success and true on failure !
+       * Hence, this wrapper to make code readable !
+       */
+      return !resolveRange(dataRangeSize);
+    }
```
or we should just change the return value in resolveRange method.

snalli replied:
Yeah i thought of it and its going to break a lot of code

snalli replied:
I can fix existing code in another PR

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+       * Update them and then compare if other immutable variables are same or not.
+       */
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      blobInfo.setLifeVersion(lifeVersion);
```
blobInfo.setLifeVersion(messageInfo.getLifeVersion()).

snalli replied:
we need to set the lifeVersion member of this class as well.

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+       * If we reach here, then the cached metadata is valid and already decrypted from a previous GET/RANGE request.
+       * Just use it directly and initiate fetching data chunks.
+       */
+      requestDataChunks();
```
can we rename this to something else, just prepareDataChunks...

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      blobInfo.setLifeVersion(lifeVersion);
+      lifeVersion = messageInfo.getLifeVersion();
+      if (!serverBlobProperties.equals(receivedBlobProperties)) {
```
we have to override BlobProperties's equals method.

snalli replied:
yes done

justinlin-linkedin commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+     * Resolve range, if needed, and request data chunks
+     * @throws IllegalArgumentException
+     */
+    void requestDataChunks() throws IllegalArgumentException {
```
can we reuse this method in FirstGetChunk.handleMetadataChunk?

snalli replied:
Yes we could but I don't want to touch that right now since its very critical to all GETs

snalli replied:
I can send out another PR to reuse this method in handleMetadataBlob

justinlin-linkedin reviewed (commented):
We should add some tests for the cached first chunk.

Arun-LinkedIn commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -140,20 +145,103 @@ class GetBlobOperation extends GetOperation {
+   */
+  boolean shouldLookupMetadataCache() {
+    return (blobMetadataCache != null &&
+        options.getBlobOptions.getRange() != null &&
```
Should we enable this for segmented GETs too (`options.getBlobOptions.hasBlobSegmentIdx()`) where user can request for segments of a stitched blob via URI like <blobId>/Segment/<SegmentNumber>? It can be useful for that feature too. Looks like you are already handling initializing data chunks for this case in `CachedFirstChunk::requestDataChunks()`

snalli replied:
I need to understand segmented GETs

snalli replied:
ok updated code to use cache for segmented GETs also

Arun-LinkedIn commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+       * Update them and then compare if other immutable variables are same or not.
+       */
+      updateTtlIfRequired(serverBlobProperties, messageInfo);
+      blobInfo.setLifeVersion(lifeVersion);
```
Should this come after line 1848 where updated life version is fetched?

snalli replied:
correct

Arun-LinkedIn commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+        chunkMetadataList = chunkMetadataList.subList(requestedSegment, requestedSegment + 1);
+        dataRangeSize = chunkMetadataList.get(0).getSize();
+      }
+      if (resolveRangeProper(dataRangeSize)) {
```
If we decide to look up cache for segmented GETs too in `shouldLookupMetadataCache`, we might need to check if this a range request by `if (options.getBlobOptions.getRange() != null) ` here.

snalli replied:
agreed, i need to read up a bit on segmented GETs

snalli replied:
For this, we only reach this code for a range request. So no need to check getRange() != null. This a policy controlled by shouldLookUpMetadataCache()

snalli replied:
ok updated code to use cache for segmented GETs also

Arun-LinkedIn commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1774,127 @@ private void onInvalidRange(Exception exception) {
+     * @throws IllegalArgumentException
+     */
+    void requestDataChunks() throws IllegalArgumentException {
+      long dataRangeSize = totalSize;
```
nit: may be rename variable `dataRangeSize` to `rangeTotalSize`

Arun-LinkedIn commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -219,6 +316,12 @@ private void onChunkOperationComplete(GetChunk chunk) {
       if (operationCallbackInvoked.compareAndSet(false, true)) {
         if (chunk.chunkBlobId.getBlobDataType() == BlobId.BlobDataType.METADATA) {
+          /**
+           * Save metadata on successful completion. Everything is decrypted at this point.
```
nit: same as above. Think it can be implementation comments (either // or /* * */) and not java doc (/** * */) comments.

snalli replied:
done

Arun-LinkedIn reviewed (commented):
Mostly looks good. Just had few comments.

snalli commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -140,20 +145,103 @@ class GetBlobOperation extends GetOperation {
+   * Unconditionally deletes blob metadata for composite blobs.
+   * If it was never saved, then we don't have to delete it.
+   * @return True if metadata was deleted successfully, else False.
+   * @param reason
```
Good point. The return val from this method is not currently used but, in general, i think its a good practice to return something assuming it may be used in the future.

snalli commented:
> We should add some tests for the cached first chunk.

yeah working on it

snalli commented:
Addressed all comments. WIP on unit tests.
LMK if I missed any comments.
The policy right now is:
1. Populate cache on each GET/RANGE/SEGMENTed GET if blobSize >= 1 TB (configurable)
2. Lookup cache only for RANGE/SEGMENT requests

Arun-LinkedIn commented on ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:
```diff
@@ -1658,6 +1751,133 @@ private void onInvalidRange(Exception exception) {
+          chunkMetadataList = compositeBlobInfo.getStoreKeysInByteRange(resolvedByteRange.getStartOffset(),
+              resolvedByteRange.getEndOffset());
+        }
+        initializeDataChunks();
```
Would you need to call `initializeDataChunks` outside the `if (options.getBlobOptions.getRange() != null)` since we also need to initialize it for segmented GETs now. But you might still need to only initialize if range resolution passed before initializing.

snalli replied:
problem with segmented get
fixBlobSizeIfRequired(blobInfo.getBlobProperties().getBlobSize(),
                  options.getBlobOptions.hasBlobSegmentIdx() ? firstChunk.getChunkMetadataList().get(0).getSize()
                      : totalSize);

This changes the size of the blob in the metadata cache and then the cached entry doesn't match blobinfo coming from backend. Ignoring SEGMENTED GET as discussed offline

snalli commented:
Fixed unittest failures
The caching policy right now is quite conservative as below. If this passes EI, then we can think of expanding the policy.

1. Populate cache on each RANGE request if blobSize >= 1 TB (configurable)
2. Lookup cache only for RANGE requests

snalli commented:
just running some tests with the cache enabled. WIP.

justinlin-linkedin reviewed (approved):
LGTM

Arun-LinkedIn reviewed (approved):
Looks good to me too
